src/server/zset_family.cc

```diff
@@ -2840,55 +2821,166 @@ void SortIfNeeded(GeoArray* ga, Sorting sorting, uint64_t count) {
+      if (shard->shard_id() == from_shard) {
+        member_score = OpScore(t->GetOpArgs(shard), key, member);
+      }
+      return OpStatus::OK;
+    };
+    cntx->transaction->Execute(std::move(cb), false);
+    auto member_sts = member_score.status();
+    if (member_sts != OpStatus::OK) {
+      cntx->transaction->Conclude();
+      switch (member_sts) {
+        case OpStatus::WRONG_TYPE:
+          return cntx->SendError(kWrongTypeErr);
+        case OpStatus::KEY_NOTFOUND:
+          return rb->StartArray(0);
+        case OpStatus::MEMBER_NOTFOUND:
+          return cntx->SendError(kMemberNotFound);
+        default:
+          return cntx->SendError(member_sts);
+      }
+    }
+    ScoreToLongLat(*member_score, shape->xy);
+  } else {
+    // verify key is valid
+    OpResult<void> result;
+    auto cb = [&](Transaction* t, EngineShard* shard) {
+      if (shard->shard_id() == from_shard) {
+        result = OpKeyExisted(t->GetOpArgs(shard), key);
+      }
+      return OpStatus::OK;
+    };
+    cntx->transaction->Execute(std::move(cb), false);
+    auto result_sts = result.status();
+    if (result_sts != OpStatus::OK) {
+      cntx->transaction->Conclude();
+      switch (result_sts) {
+        case OpStatus::WRONG_TYPE:
+          return cntx->SendError(kWrongTypeErr);
+        case OpStatus::KEY_NOTFOUND:
+          return rb->StartArray(0);
+        default:
+          return cntx->SendError(result_sts);
+      }
+    }
+  }
+  DCHECK(shape->xy[0] >= -180.0 && shape->xy[0] <= 180.0);
+  DCHECK(shape->xy[1] >= -90.0 && shape->xy[1] <= 90.0);
+
+  // query
   GeoHashRadius georadius = geohashCalculateAreasByShapeWGS84(shape);
   GeoArray ga;
-  if (!MembersOfAllNeighbors(cntx, key, georadius, shape_ref, &ga,
-                             geo_ops.any ? geo_ops.count : 0)) {
-    return;
+  auto range_specs = GetGeoRangeSpec(georadius);
+  // get all the matching members and add them to the potential result list
+  vector<OpResult<vector<ScoredArray>>> result_arrays;
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    auto res_it = OpRanges(range_specs, t->GetOpArgs(shard), key);
+    if (res_it) {
+      result_arrays.emplace_back(res_it);
+    }
+    return OpStatus::OK;
+  };
+  cntx->transaction->Execute(std::move(cb), false);
+
+  // filter potential result list
+  double xy[2];
+  double distance;
+  unsigned long limit = geo_ops.any ? geo_ops.count : 0;
+  for (auto& result_array : result_arrays) {
+    for (auto& arr : *result_array) {
+      for (auto& p : arr) {
+        if (geoWithinShape(shape, p.second, xy, &distance) == 0) {
+          ga.emplace_back(xy[0], xy[1], distance, p.second, p.first);
+          if (limit > 0 && ga.size() >= limit)
+            break;
+        }
+      }
+    }
   }
 
-  auto* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
   // if no matching results, the user gets an empty reply.
   if (ga.empty()) {
     rb->SendNull();
+    cntx->transaction->Conclude();
```

Comment:
what's the expected behaviour for  no matched result together with STORE option? do we have a test for that? are we compliant with redis behavior in this case?